keymanapp/keyman
# fix(web): maps touch-layout chiral alt, ctrl to non-chiral when non-chiral keyboard is active

Fixes #6792.

Up until now, KeymanWeb hasn't bothered comparing a keyboard's chirality with the corresponding OSK's key "layer" property - it's simply assumed consistency between the two.  Should a keyboard developer specify a chiral "layer" property when the keyboard itself isn't actually chiral, that unchecked assumption would cause the key not to match in the expected way.  For the price of a little extra overhead, we can remove that assumption.

Note that there is no matching keyboard for the user test; instead, I've added numerous unit tests to handle the important part of the changes for the original issue.  However, it's still best to test other keyboards to make sure that others aren't affected.

## User Testing

### Tests

- TEST_NORMAL_CHIRALITY:  Using the "Chirality testing/bootstrapping" test page...
    - Test by using Chrome's remote-device emulation.
    - Use a longpress-shift and select the "LAlt" layer, then type at least 3 keys with the OSK.
    - Resulting text should perfectly match the keycaps for the keys you typed.
    - Then, tap shift and then longpress-shift again to select the "Shift + RAlt" layer, then type at least 3 keys with the OSK.
    - Resulting text should, again, perfectly match the keycaps for the keys you typed.

- TEST_CHIRALITY_HARDWARE:  Using the "Chirality testing/bootstrapping" test page...
    - Without remote-device emulation in Chrome, but with a hardware keyboard that has two Alt keys...
    - Hold LAlt.  The desktop OSK should change layer and show new keycaps.
    - Type at least three keys while holding LAlt.  The resulting text should perfectly match the keycaps for the keys you typed.
    - Then, hold Shift + RAlt.  The desktop OSK should change layer and show new keycaps. 
    - Again, type at least 3 keys with the OSK.  The resulting text should perfectly match the keycaps for the keys you typed.
    - Now, hold Shift + LAlt + LCtrl.  The desktop OSK should show the same keycaps that it did for Shift + RAlt.

- TEST_NORMAL_KEYBOARD_OSK:  Using the "Test unminified Keymanweb" test page...
    - Test by using Chrome's remote-device emulation.
    - Select the "French" keyboard.
    - Type at least two keys on the base layer and confirm they act as expected.
    - Then, use a longpress-shift to select the Alt+Ctrl layer, then type `[` and `]`, confirming they act as expected.
    - Then, type the `~` key.  Nothing should emit.
    - Then, drop to the base layer, then type an `a`.  An `ã` should result.

- TEST_NORMAL_KEYBOARD_HARDWARE:  Using the "Test unminified Keymanweb" test page...
    - Without remote-device emulation in Chrome, but with a hardware keyboard that has two Alt keys...
    - Select the "French" keyboard.
    - Type at least two unmodified (no Shift, Ctrl, Alt...) _with your hardware keyboard_ and confirm they act as expected.
    - Then, hold Alt + Ctrl and type your hardware `5` and `-` keys - they should produce `[` and `]` respectively, regardless of which Alt and Ctrl key you use.
    - Then, type Alt + Ctrl + `2` key.  Nothing should emit.
    - Then, release Alt + Ctrl, then type an `a`.  An `ã` should result.
